src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingSubmissionService.java

```diff
@@ -165,14 +173,26 @@ public ProgrammingSubmission notifyPush(Long participationId, Object requestBody
 
         programmingSubmission = new ProgrammingSubmission();
         programmingSubmission.setCommitHash(commit.getCommitHash());
-        log.info("create new programmingSubmission with commitHash: " + commit.getCommitHash() + " for participation " + participationId);
+        log.info("Create new programmingSubmission with commitHash: " + commit.getCommitHash() + " for participation " + participationId);
 
         programmingSubmission.setSubmitted(true);
         programmingSubmission.setSubmissionDate(ZonedDateTime.now());
         programmingSubmission.setType(SubmissionType.MANUAL);
-
         programmingExerciseParticipation.addSubmission(programmingSubmission);
 
+        // Students are not allowed to submit a programming exercise after the exam due date, if this happens we set the Submission to ILLEGAL
+        if (isExamExercise && programmingExerciseParticipation instanceof ProgrammingExerciseStudentParticipation) {
+            var optionalStudent = ((ProgrammingExerciseStudentParticipation) programmingExerciseParticipation).getStudent();
```

Comment:
can you move the content of the whole if statement into a separate method please? Otherwise notifyPush gets too complex

Reply:
done in 2aef655083373af302f175c973b2e466b5c230a1